lib/common/socket/evloop.c.h

```diff
@@ -530,9 +530,11 @@ static void run_socket(struct st_h2o_evloop_socket_t *sock)
             int so_err = 0;
             socklen_t l = sizeof(so_err);
             so_err = 0;
-            if (getsockopt(sock->fd, SOL_SOCKET, SO_ERROR, &so_err, &l) != 0 || so_err != 0) {
-                /* FIXME lookup the error table */
+            if (getsockopt(sock->fd, SOL_SOCKET, SO_ERROR, &so_err, &l) != 0) {
                 err = h2o_socket_error_conn_fail;
+            } else if (so_err != 0) {
+                err = h2o_socket_error_conn_fail;
+                errno = so_err;
```

Comment:
I do not think we need to set `errno`, due to the same reason as above.

If we want to notify a fine grained error from here, it should be done by adding more string constants to `h2o_socket_error_*` and setting one of them in `err`.

Reply:
Guess 'h2o_socket_error_conn_fail' is enough for upper layer or the user.and there are many reasons for connection failure. So errno is just a hint for the reason.just like Read/write error.
If not set here, the reason can be retrieved by the user manually.

Reply:
Well, the thing is that an application cannot tell if `errno` reflects an error of the socket that it is dealing with. We do not have such a contract (that `error` reflects the state of the socket when some callbacks are called), and we cannot have such a contract especially when libuv is used as a event loop.

Reply:
Thanks for your patient explanation.
`getsockopt(SO_ERROR)` should be used by application when needed to obtain the socket error.
guess this also applied to libuv.